File: BackEnd/TypeScript/nestshop_typeorm/src/friend/friend.controller.ts
import { Controller, Delete, Get, Param, ParseIntPipe, Patch, Post, Query } from '@nestjs/common';
import { ApiBearerAuth, ApiOperation, ApiTags } from '@nestjs/swagger';
import { CurrentUser, JwtPayload } from '../common/decorators/current-user.decorator';
import { FriendPaginationQueryDto } from './dto/friend-pagination-query.dto';
import { FriendService } from './friend.service';

@ApiTags('Friend')
@ApiBearerAuth()
@Controller('friends')
export class FriendController {
  constructor(private readonly friendService: FriendService) {}

  @Post('request/:userId')
  @ApiOperation({ summary: '친구 신청' })
  requestFriend(@CurrentUser() user: JwtPayload, @Param('userId', ParseIntPipe) userId: number) {
    return this.friendService.requestFriend(user.sub, userId);
  }

  @Patch('request/:friendshipId/accept')
  @ApiOperation({ summary: '친구 신청 수락' })
  acceptRequest(
    @CurrentUser() user: JwtPayload,
    @Param('friendshipId', ParseIntPipe) friendshipId: number,
  ) {
    return this.friendService.acceptRequest(user.sub, friendshipId);
  }

  @Patch('request/:friendshipId/reject')
  @ApiOperation({ summary: '친구 신청 거절' })
  rejectRequest(
    @CurrentUser() user: JwtPayload,
    @Param('friendshipId', ParseIntPipe) friendshipId: number,
  ) {
    return this.friendService.rejectRequest(user.sub, friendshipId);
  }

  @Get()
  @ApiOperation({ summary: '내 친구 목록 조회' })
  getFriends(@CurrentUser() user: JwtPayload, @Query() query: FriendPaginationQueryDto) {
    return this.friendService.getFriends(user.sub, query);
  }

  @Get('requests/received')
  @ApiOperation({ summary: '받은 친구 요청 목록' })
  getReceivedRequests(@CurrentUser() user: JwtPayload, @Query() query: FriendPaginationQueryDto) {
    return this.friendService.getReceivedRequests(user.sub, query);
  }

  @Get('requests/sent')
  @ApiOperation({ summary: '보낸 친구 요청 목록' })
  getSentRequests(@CurrentUser() user: JwtPayload, @Query() query: FriendPaginationQueryDto) {
    return this.friendService.getSentRequests(user.sub, query);
  }

  @Get('feed')
  @ApiOperation({ summary: '친구 활동 피드 조회' })
  getFeed(@CurrentUser() user: JwtPayload, @Query() query: FriendPaginationQueryDto) {
    return this.friendService.getFeed(user.sub, query);
  }

  @Post('block/:userId')
  @ApiOperation({ summary: '유저 차단' })
  blockUser(@CurrentUser() user: JwtPayload, @Param('userId', ParseIntPipe) userId: number) {
    return this.friendService.blockUser(user.sub, userId);
  }

  @Delete('block/:userId')
  @ApiOperation({ summary: '유저 차단 해제' })
  unblockUser(@CurrentUser() user: JwtPayload, @Param('userId', ParseIntPipe) userId: number) {
    return this.friendService.unblockUser(user.sub, userId);
  }

  @Delete(':userId')
  @ApiOperation({ summary: '친구 삭제' })
  removeFriend(@CurrentUser() user: JwtPayload, @Param('userId', ParseIntPipe) userId: number) {
    return this.friendService.removeFriend(user.sub, userId);
  }
}

File: BackEnd/Kotlin/kotlin-ktor-gradle-exposeddao-postgresql/src/main/kotlin/com/pbshop/kotlin/ktor/gradle/exposeddao/postgresql/friend/FriendController.kt
package com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.friend

import com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.common.pbActor
import com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.common.respondStub
import com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.security.PbRole
import com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.security.requireAnyRole
import io.ktor.server.application.call
import io.ktor.server.routing.Route
import io.ktor.server.routing.delete
import io.ktor.server.routing.get
import io.ktor.server.routing.patch
import io.ktor.server.routing.post

class FriendController(
    private val service: FriendService,
) {
    fun Route.register() {
        post("/friends/request/{userId}") {
            call.requireAnyRole(PbRole.USER)
            call.respondStub(service.requestFriend(call.currentUserId(), call.pathInt("userId")))
        }
        patch("/friends/request/{friendshipId}/accept") {
            call.requireAnyRole(PbRole.USER)
            call.respondStub(service.acceptRequest(call.currentUserId(), call.pathInt("friendshipId")))
        }
        patch("/friends/request/{friendshipId}/reject") {
            call.requireAnyRole(PbRole.USER)
            call.respondStub(service.rejectRequest(call.currentUserId(), call.pathInt("friendshipId")))
        }
        get("/friends") {
            call.requireAnyRole(PbRole.USER)
            call.respondStub(
                service.listFriends(
                    userId = call.currentUserId(),
                    page = call.request.queryParameters["page"]?.toIntOrNull() ?: 1,
                    limit = call.request.queryParameters["limit"]?.toIntOrNull() ?: 20,
                ),
            )
        }
        get("/friends/requests/received") {
            call.requireAnyRole(PbRole.USER)
            call.respondStub(
                service.receivedRequests(
                    userId = call.currentUserId(),
                    page = call.request.queryParameters["page"]?.toIntOrNull() ?: 1,
                    limit = call.request.queryParameters["limit"]?.toIntOrNull() ?: 20,
                ),
            )
        }
        get("/friends/requests/sent") {
            call.requireAnyRole(PbRole.USER)
            call.respondStub(
                service.sentRequests(
                    userId = call.currentUserId(),
                    page = call.request.queryParameters["page"]?.toIntOrNull() ?: 1,
                    limit = call.request.queryParameters["limit"]?.toIntOrNull() ?: 20,
                ),
            )
        }
        get("/friends/feed") {
            call.requireAnyRole(PbRole.USER)
            call.respondStub(
                service.feed(
                    userId = call.currentUserId(),
                    page = call.request.queryParameters["page"]?.toIntOrNull() ?: 1,
                    limit = call.request.queryParameters["limit"]?.toIntOrNull() ?: 20,
                ),
            )
        }
        post("/friends/block/{userId}") {
            call.requireAnyRole(PbRole.USER)
            call.respondStub(service.blockUser(call.currentUserId(), call.pathInt("userId")))
        }
        delete("/friends/block/{userId}") {
            call.requireAnyRole(PbRole.USER)
            call.respondStub(service.unblockUser(call.currentUserId(), call.pathInt("userId")))
        }
        delete("/friends/{userId}") {
            call.requireAnyRole(PbRole.USER)
            call.respondStub(service.removeFriend(call.currentUserId(), call.pathInt("userId")))
        }
    }

    private fun io.ktor.server.application.ApplicationCall.currentUserId(): Int = pbActor().userId ?: 0

    private fun io.ktor.server.application.ApplicationCall.pathInt(name: String): Int = parameters[name]?.toIntOrNull() ?: 0
}
